PassiveSonarOS™

DSsoto opened the pull request:
This is a complete revamp of the passive sonar packages of the past. It is well documented, with a wiki page to come soon. The net result is a new passive sonar driver that handles everything from the acquisition of the signals via serial, the estimation of time deltas, the interface for other nodes to query the heading towards the pinger or to estimate the full position of the pinger, and the visualization of all results in RVIZ.

The code has been restructured such that all configuration can be done be the rosparam tag in a launch and some yaml. 

Also the package will work without modifications to source code for an arbitrary number of receivers in an arbitrary arrangement, as long as there is enough information to solve the multilateration problem. 

New functionality was added to the multilateration python pkg, most prominently functions to calculate the least-squares intersection of a set of (2D or 3D) lines.

The great_merge_purgatory in mil_vision was removed as promised. CameraLidarTransformer was kept, while the other files were removed.

This will all be tested at the pool on April 3rd, so we should be confident merging it after that.

DSsoto committed:
PASSIVE SONAR: add 3D line intersection estimation

Also refactor Navigator's old pinger finder node to work in 3d and do code cleanup.

DSsoto committed:
PASSIVE SONAR: add updated Jake board driver code

Updates:
* New iteration of board has a few new param values (baud, serial port).
* Jake has modified some of the serial communication code.

DSsoto committed:
PASSIVE SONAR: rename sonar_driver.py to mil_passsive_sonar.py

DSsoto committed:
PASSIVE SONAR: add example launch file

DSsoto committed:
PASSIVE SONAR: add ros srvs used for driver

DSsoto committed:
PASSIVE SONAR: overhaul of sonar driver package

Multilateration python pkg:
* Functions for finding the LS intersection of 2D or 3D lines were added
* A function to calculate the maximum possible dtoa for a pair of receivers was added
* A function to find the delay between two nearly identical signals was added
* Added docstrings for all functions

Passive Sonar driver node:
* All configurale elements wwere made rosparams, an example launch is given
* Magic constants have been removed
* Repetitive code has been refactored
* Added visualization of headings and position estimate in RVIZ
* Added service to reset position estimate
* Removed dependence on SubjuGator msgs
* Added standalone function to receive signals from a serial port
* Added function to consolidate the loading of rosparams
* Added docstrings for all functions
* Added timeout for listening to receiver board
* Restructured code to work with an arbitrary number of receivers in an
  arbitrary arrangement
* Remove dependence on sub8_alarms

DSsoto committed:
PASSIVE SONAR: add changes from test day

mattlangford commented on drivers/mil_passive_sonar/mil_passive_sonar/passive_sonar_driver.py:
```diff
@@ -0,0 +1,315 @@
+            self.ser = serial.Serial(port=self.port, baudrate=self.baud, timeout=self.request_timeout)
+            self.ser.flushInput()        
+        except Exception, e:
+            print "\x1b[31mSonar serial connection error:\n\t", e, "\x1b[0m"
```
Use ros logging here

mattlangford commented on drivers/mil_passive_sonar/mil_passive_sonar/passive_sonar_driver.py:
```diff
@@ -0,0 +1,315 @@
+        self.tf_listener = tf.TransformListener()
+        self.multilaterator = Multilaterator(self.receiver_locations, self.c, self.method)
+
+        self.rviz_pub = rospy.Publisher("/visualization_marker", Marker, queue_size=10)
```
Consider a more unique name for vidualizations

mattlangford commented on drivers/mil_passive_sonar/mil_passive_sonar/passive_sonar_driver.py:
```diff
@@ -0,0 +1,315 @@
+        self.c = load('c')
+        self.port = load('port')
+        self.baud = load('baud')
+	self.TX_REQUEST = load('tx_request_code')
```
Does this work with the broken spacing?

mattlangford commented on drivers/mil_passive_sonar/mil_passive_sonar/passive_sonar_driver.py:
```diff
@@ -0,0 +1,315 @@
+        Uses a buffer of prior observations to estimate the position of the pinger as the intersection
+        of a set of 3d lines in the least-squares sense.
+        '''
+        assert len(self.heading_start) > 1
```
Will this assertion cause the script to crash? If so it'd probably be favorable to just return a blank response with 0 observations

mattlangford commented on drivers/mil_passive_sonar/mil_passive_sonar/passive_sonar_driver.py:
```diff
@@ -0,0 +1,315 @@
+        Clears all the heading and amplitude buffers and makes the position estimate NaN
+        '''
+	self.heading_start = np.empty((0, 3), float)
+	self.heading_end = np.empty((0, 3), float)
```
There's more spacing issues here. I know python will break under certain conditions if the spacing is off by too much.

mattlangford commented on utils/mil_msgs/srv/GetPulseHeading.srv:
```diff
@@ -0,0 +1,7 @@
+
+# Heading unit vector in body frame
+Header header
+geometry_msgs/Vector3 heading
```
Might be good to add a `is_valid` flag here

mattlangford reviewed (commented):
I might have some more comments when I'm not on mobile. Let me know if those spacing issues actually cause problems.

Will leave this as a comment pending a computer read though

DSsoto committed:
PASSIVE SONAR: remove dependence on mil_msgs

All srvs were moved to the inside of the pkg. This will allow us to open source
the pkg with minimal dependencies.

Added descriptive comments to the .srv files.

DSsoto committed:
PASSIVE SONAR: rename mil_passive_sonar python module

Renamed to passive_sonar_driver to avoid problem where python would not be able
to find the mil_passive_sonar.srv module due to name conflict.

DSsoto committed:
PASSIVE SONAR: streamline loading of rosparams

PASSIVE SONAR: streamline declaration of services

PASSIVE SONAR: separate serial comms code from driver class

DSsoto committed:
PASSIVE_SONAR: create base class for input to sonar driver

PASSIVE SONAR: create serial input class

DSsoto committed:
PASSIVE SONAR: instantiate input source class based on input mode

PASSIVE SONAR: simplify input source interface

Only one method is used now to get input

PASSIVE SONAR: switch to tf2

PASSIVE SONAR: add function to record input data

PASSIVE SONAR: add service handler to set bagging mode

DSsoto committed:
PASSIVE SONAR: fix bug in calculation of time deltas

DSsoto committed:
PASSIVE SONAR: add function to visualize dsp operations

DSsoto committed:
PASSIVE SONAR: publish dsp plot images

DSsoto committed:
PASSIVE SONAR: rename Receiver Array interface base class

DSsoto committed:
PASSIVE SONAR: add ReceiverArrayInterface for bagged data

DSsoto committed:
PASSIVE SONAR: implement saving data to bag

DSsoto committed:
PASSIVE SONAR: make tf getter a member of serial interface

mattlangford commented on drivers/mil_passive_sonar/multilateration/multilateration.py:
```diff
@@ -1,73 +1,184 @@
+    '''
+    Find the intersection of lines in the least-squares sense.
+    start - Nx3 numpy array of start points
+    end - Nx3 numpy array of end points
```
Could you explain this process or link to an explanation here

mattlangford commented on drivers/mil_passive_sonar/multilateration/multilateration.py:
```diff
@@ -1,73 +1,184 @@
         print "\x1b[32mSpeed of Sound (c):", self.c, "millimeter/microsecond\x1b[0m"
 
-    def getPulseLocation(self, timestamps, method=None):
+    def getPulseLocation(self, dtoa, method=None):
```
Should_be_consistent_with_function_naming_scheme
Also should this always return something, you could get keyboard interrupted and not return anything (though anytime you get keyboard interrupted you probably want to stop the program anyway)

DSsoto replied:
Fixed, and always return, and better to throw exception than assert.

mattlangford commented on drivers/mil_passive_sonar/multilateration/multilateration.py:
```diff
@@ -106,94 +217,96 @@ def get_B(delta):
-            xi = self.hydrophone_locations[i][0]
-            yi = self.hydrophone_locations[i][1]
-            zi = self.hydrophone_locations[i][2]
+        d0 = np.sqrt((x)**2 + (x)**2 + (x)**2)
```
Should this be `x`, `y`, and `z` instead of `x`, `x`, `x`

DSsoto replied:
Yes

mattlangford commented on drivers/mil_passive_sonar/multilateration/multilateration.py:
```diff
@@ -106,94 +217,96 @@ def get_B(delta):
+            xi = self.receiver_locations[i, 0]
+            yi = self.receiver_locations[i, 1]
+            zi = self.receiver_locations[i, 2]
             di = np.sqrt((xi - x)**2 + (yi - x)**2 + (zi - x)**2)
```
Same with the `x`, `y`, `z`? If so this is the same as `np.linalg.norm` which would be more faster

DSsoto replied:
Yeah this is code from my pleb tier python days. I need to fix it up and add some unit tests.

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+# GLOBALS
+lock = threading.Lock()  # prevent multiple threads simultaneously using a serial port
+
+def thread_lock(lock):  # decorator
```
There's a `mil_tool` for this in `threading_helpers`. I think it does the same thing but it would reduce duplicated code.

DSsoto replied:
I know but if we contribute this package we wouldn't want it to depend on `mil_tools`.

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+        return locked_function
+    return lock_thread
+
+class ReceiverArrayInterface(object):
```
Good idea for playing back from bags vs reading from serial

DSsoto replied:
Haha thanks  :sunglasses:
Also going to add a Gazebo interface

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+        load = lambda prop: setattr(self, prop, rospy.get_param('passive_sonar/' + prop))
+        [load(x) for x in param_names]
+
+        self.tf2_buf = tf2_ros.Buffer()
```
The future

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+            start of a transmission
+        '''
+        self.ser.flushInput()
+        try:
```
Wouldn't you want errors here to propagate up to the try statement in `intput_request` so that it won't do calculations with bad data?

DSsoto replied:
Yeah, you're right.

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+                    pass
+                self.signals[channel, i] = float(self.ser.read(self.scalar_size)) - self.signal_bias
+
+    def getReceiverPose(time, receiver_array_frame, locating_frame):
```
~~camal case~~

DSsoto replied:
fixed

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+
+        for channel in range(self.num_receivers):
+            for i in range(self.signal_size):
+                while self.ser.inWaiting() < self.scalar_size:
```
Should probably add some ros protection here, and maybe sleep in the loop instead of passing.

DSsoto replied:
Added sleeping. What kind of ros protection? I don't see the necessity here.

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+
+        returns: 2D numpy array with <num_signals> rows and <signal_size> columns
+        '''
+        self.signals = np.full((self.num_receivers, self.signal_size), self.signal_bias, dtype=float)
```
This would be a good place for [a custom np.dtype](https://docs.scipy.org/doc/numpy/reference/arrays.dtypes.html).

DSsoto replied:
Why do you think so? I'm just converting an ascii string  to a float, and all other code here only cares about the float representation.

mattlangford replied:
The dtype is composed of floats. You'd have a dtype that looks like this: `[('x', float), ('y', float), ('z', float), ('t', float)]`. Creating it would be `self.signals = np.zeros(self.num_receivers, dtype=my_dtype)`. Then filling it in would be almost the same. But using the data would make more sense. Instead of just indices`self.signals[0][0] * self.signals[1][2] + ...` you can reference fields by name `self.signals[0]['x'] * self.signals[1]['z'] + ...`. It's something I've started doing as to lessen the number of random indexing I'm doing and make it more clear to the reader of the code what's going on.

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+            for i in range(self.signal_size):
+                while self.ser.inWaiting() < self.scalar_size:
+                    pass
+                self.signals[channel, i] = float(self.ser.read(self.scalar_size)) - self.signal_bias
```
Does the board send 32 bit floats or 64 bit floats, I think python does everything in 64? It seems like this could cause confusion or problems at some point. It might be good to use `np.float64(...)` to make it clear what kind of data it's actually reading.

Also wouldn't it be good to do some CRC checking?

DSsoto replied:
It sends 5 decimal ascii characters. I added some validity checking. CRC would be nice but that's up to Jake. For other users they can pass in their own object that decodes the serial bytes.

mattlangford replied:
That seems strange, why not full precision floating point numbers?

DSsoto replied:
Yeah it's a bit strange, its a Jake thing, idk.

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+                                              timeout=rospy.Duration(0.20))
+            q = T.transform.rotation
+            t = T.transform.translation
+            rot = transformations.quaternion_matrix([q.w, q.x, q.y, q.z])
```
Idea for a helper function: tf(2) response to numpy position, rotation (3x3)

DSsoto replied:
That's a good idea, I explicitly didn't do that or use other helpers to avoid the dependency on other mil pkgs.

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+        self.load_params()
+        self.bagging = False
+
+        # TODO: update to use ros_alarms
```
That's right

DSsoto replied:
Eyyyyy....

DSsoto replied:
Although it would create a dependence on another mil pkg...

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+        sonar board and calculating headings towards an active pinger.
+
+        These parameters are descrived in detail in the Passive Sonar page of the mil_common wiki.
+        TODO: copy url here
```
^

DSsoto replied:
I need to create it first lol

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+            raise IOError('A required rosparam was not declared: ' + str(e))
+
+        self.receiver_count = len(self.receiver_locations) + 1
+        self.receiver_locations = np.array(  # dictionary to numpy array
```
I think with custom dtypes you wouldn't have to do this, you could save the dictionary as an indexable numpy array where `self.receiver_locations['x']` gives you all the x values just like `self.receiever_locations[:, 0]` would. Food for thought

DSsoto replied:
I'll have to use those in the future, that sounds pretty useful. However, since this a one-off thing, I'll leave it as is.

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+        except KeyError as e:
+            raise IOError('A required rosparam was not declared: ' + str(e))
+
+        self.receiver_count = len(self.receiver_locations) + 1
```
Wouldn't having to specify all `receiver_locations` make more sense, does one of the hydrophones _have_ to be the origin? (Assuming the +1 is because you don't have to specify the (0, 0, 0) center hydrophone which may be bad assumption)

DSsoto replied:
All of this is based around the idea that there is a special receiver that all other time delta's will be measured with respect to. Therefore it should also be the basis for the coordinate system of the array. I think this is pretty reasonable and simplifies things. (You are right about the +1)

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+                      'save_bag'                 : SaveBag,
+                      'set_frequency'            : SetFrequency
+                  }
+       [rospy.Service('passive_sonar/' + s[0], s[1], getattr(self, s[0])) for s in services.items()]
```
Do the services stay alive when you don't assign them to a variable? I feel like I've had issues where the service wouldn't be available in the past when I don't set the instantiation to a variable - thought I could have been a fool in those days.

DSsoto replied:
You were a fool in those days.

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+
+            # Poll input source until it reports it is ready
+            self.input_source.input_request()
+            while not self.input_source.ready:
```
`and not rospy.is_shutdown():`
Even though you time out below, I feel like it'd be safer to have the control-c option in case of anomaly.

DSsoto replied:
Sure

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+            signals, p0, R = self.input_source.get_input()
+
+            # Carry out multilateration to get heading to pinger
+            heading = self.multilaterator.getPulseLocation(self.get_dtoa(signals))
```
This could be all zeros and throw an error, but it wouldn't be obvious why the pulse location wasn't valid when that divide by 0 error get's thrown. It would probably be good to add error logs in the `Multilaterator` class so you know why the estimate isn't valid.

DSsoto replied:
A heading will never be all zeros, but valid point about error logging in that class I'll check up on it.

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+        '''
+        Clears all the heading and amplitude buffers and makes the position estimate NaN
+        '''
+	self.heading_start = np.empty((0, 3), float)
```
Spacing troubles here, this doesn't cause crashing?

DSsoto replied:
I copied a tab by accident. Fixed.

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+        self.pinger_position = np.array([np.NaN, np.NaN, np.NaN])
+        return {}
+
+    def start_bagging(self, req):
```
I would tend to think that this shouldn't be called bagging since it's not a rosbag and can cause confusion. Maybe call it data logging or something?

DSsoto replied:
Ok, called it logging.

mattlangford replied:
Make sure to change it everywhere

mattlangford commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+        marker.type = marker.SPHERE
+        marker.action = marker.ADD
+        marker.scale.x = 0.2
+        np.clip(bgra, 0.0, 1.0)
```
This line doesn't mutate the `bgra` values, does it? `bgra = np.clip(bgra, ...)`

DSsoto replied:
Good catch

mattlangford commented on drivers/mil_passive_sonar/srv/EstimatePingerPosition.srv:
```diff
@@ -0,0 +1,13 @@
+Header header
+
+# Number of heading observations used to calculate position estimate
+int32  num_headings
```
`uint32`, gotta save room for those other 2147483647+/-1 obersvations

DSsoto replied:
Done

mattlangford reviewed (changes requested):
I want pizza

I look forward to your responses.

DSsoto committed:
PASSIVE SONAR: address PR17 comments

* Rename bagging to logging to avoid confusion
* Remove dependency on sklearn
* Replace asserions with exceptions
* Provide link to explanation of LS intersection code
* Fix naming inconsistencies
* Fix bug in one of the solver cost fucntions
* Add correction of non-decimal ascii characters (serial)
* Add ros shutdown checking to serial comms code
* Clean up matplotlib plots

kev-the-dev commented on drivers/mil_passive_sonar/multilateration/multilateration.py:
```diff
@@ -1,73 +1,184 @@
 from scipy import optimize
 from itertools import combinations
 
 from sub8_msgs.srv import Sonar, SonarResponse
```
Should not depend on sub8, navigator

kev-the-dev commented on drivers/mil_passive_sonar/nodes/sonar_test.py:
```diff
@@ -10,6 +10,9 @@
 import sys
 
 
+# This should be turned into a unit test
```
Should be done before merging, use rostest framework

DSsoto replied:
Agreed, I separated the sonar stuff from the great merge stuff and will add tests before merging.

kev-the-dev commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,590 @@
+       rospy.logerr('The param passive_sonar/input_mode was not set, defaulting to \
+                    \'serial\'. (' + str(e) + ')')
+
+    ping_ping_motherfucker = PassiveSonar(input_mode=mode)
```
We're trying to move away from potty mouth language in MIL

DSsoto replied:
But its not even part of Carlin's list. :(

kev-the-dev reviewed (approved):
I'd be cool if we separate your great merge refactors from the sonar os while you're still working on it if the git operation isn't too hard

DSsoto committed:
MIL MSGS: remove unused dependencies

Removed depenencies to interactive_markers, tf, and eigen

sentree reviewed (changes requested):
Everything in drivers/mil_passive_sonar is currently excluded from formatting checks. Before this is merged, the check should be enabled for that directory and everything in it should be properly formatted.

kev-the-dev commented on drivers/mil_passive_sonar/passive_sonar_driver/passive_sonar_driver.py:
```diff
@@ -0,0 +1,613 @@
+        of a set of 3d lines in the least-squares sense.
+        '''
+        if len(self.heading_start) > 1:
+            raise RuntimeError(
```
I'd be better to add a success attribute to the service response and return success = False instead of crashing if this is requested without headings found

kev-the-dev reviewed (changes requested):
I encountered a scenerio when running the driver in sim in which headings were added despite being NaN. Here's a diff which fixed it:
```
diff --git a/drivers/mil_passive_sonar/passive_sonar_driver/driver.py b/drivers/mil_passive_sonar/passive_sonar_driver/driver.py
index 31942ab..b22a523 100755
--- a/drivers/mil_passive_sonar/passive_sonar_driver/driver.py
+++ b/drivers/mil_passive_sonar/passive_sonar_driver/driver.py
@@ -269,7 +269,7 @@ class PassiveSonar(object):
         try:
             variance = np.var(np.array([s.samples for s in signals]).flatten())
             # if variance > self.min_variance:  # Volume threshold
-            if np.all(np.isfinite(variance)):  # checks for NAN or inf / -inf
+            if np.all(np.isfinite(variance)) and np.all(np.isfinite(heading)):  # checks for NAN or inf / -inf
                 delta = R.dot(heading)
                 p1 = p0 + delta
 ```
